File: cpp/leetcode/206.cpp
class Solution {
public:
    ListNode *reverseList(ListNode *head) {
        if (!head || !head->next) {
            return head;
        }
        ListNode *nhead = NULL;
        while (head) {
            ListNode *tmp = head->next;
            head->next = nhead;
            nhead = head;
            head = tmp;
        }
        return nhead;
    }
};

class Solution {
public:
    ListNode *reverseList(ListNode *head) {
        if (!head || !head->next) {
            return head;
        }
        ListNode *nhead = reverseList(head->next);
        head->next->next = head;
        head->next = NULL; // set the end of list to null
        return nhead;
    }
};

File: python/leetcode/206.py
class Solution(object):
    def reverseList(self, head):
        if not head or not head.next:
            return head
        tmp = head.next
        nhead = self.reverseList(tmp)
        tmp.next = head
        head.next = None
        return nhead
        

class Solution(object):
    def reverseList(self, head):
        if not head or not head.next: 
            return head 
        nhead = None
        while head:
            tmp = head.next
            head.next = nhead
            nhead = head
            head = tmp
        return nhead